domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,212 @@
+            )
+          }
+        }
+      }
+    }.also {
+      CoroutineScope(backgroundCoroutineDispatcher).launch {
+        sendTakeSnapshotCommand(switchIconificationCount)
+      }
+    }
+  }
+
+  @ObsoleteCoroutinesApi
+  private fun sendSwitchIconificationCommand(newIconification: AppIconification) {
+    commandQueue.offer(CommandMessage.SwitchIconification(newIconification)).apply {
+      if (!this) {
+        val exception = IllegalStateException("Iconification switching failed")
+        consoleLogger.e(
+          "CpuPerformanceSnapshotter",
+          "Failure while switching AppIconification.",
+          exception
+        )
+        exceptionLogger.logException(exception)
+      }
+    }
+  }
+
+  @ObsoleteCoroutinesApi
+  private suspend fun sendScheduleTakeSnapshotCommand(
+    currentIconification: AppIconification,
+    switchId: Int
+  ) {
+    commandQueue.send(CommandMessage.ScheduleTakeSnapshot(currentIconification, switchId))
+  }
+
+  @ObsoleteCoroutinesApi
+  private suspend fun sendTakeSnapshotCommand(switchId: Int) {
+    commandQueue.send(CommandMessage.TakeSnapshot(switchId))
+  }
+
+  @ObsoleteCoroutinesApi
+  private suspend fun sendLogSnapshotDiffCommand(first: Snapshot, second: Snapshot) {
+    commandQueue.send(CommandMessage.LogSnapshotDiff(first, second))
+  }
+
+  @ObsoleteCoroutinesApi
+  private fun scheduleTakeSnapshot(currentIconification: AppIconification, switchId: Int) {
+    val delayMs = when (currentIconification) {
+      AppIconification.APP_IN_FOREGROUND -> foregroundCpuLoggingTimePeriod
+      AppIconification.APP_IN_BACKGROUND -> backgroundCpuLoggingTimePeriod
+    }
+    CoroutineScope(backgroundCoroutineDispatcher).launch {
+      delay(delayMs)
+      sendTakeSnapshotCommand(switchId)
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          "PerformanceMetricsController",
+          "Failed to remove metric log.",
+          it
+        )
+      }
+    }
+  }
+
+  private fun computeSnapshotAtCurrentTime(iconification: AppIconification): Snapshot {
+    return Snapshot(
+      appTimeMillis = oppiaClock.getCurrentTimeMs(),
+      cpuTimeMillis = Process.getElapsedCpuTime(),
+      numCores = getNumberOfOnlineCores(),
+      iconification = iconification
+    )
+  }
+
+  private fun getRelativeCpuUsage(firstSnapshot: Snapshot, secondSnapshot: Snapshot): Double {
+    val deltaCpuTimeMs = secondSnapshot.cpuTimeMillis - firstSnapshot.cpuTimeMillis
+    val deltaProcessTimeMs = secondSnapshot.appTimeMillis - firstSnapshot.appTimeMillis
+    val numberOfCores = (secondSnapshot.numCores + firstSnapshot.numCores) / 2
+    return deltaCpuTimeMs.toDouble() / (deltaProcessTimeMs.toDouble() * numberOfCores)
+  }
+
+  private fun getNumberOfOnlineCores(): Int {
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      Os.sysconf(OsConstants._SC_NPROCESSORS_ONLN).toInt()
```

Comment:
Have you confirmed that this returns the values expected?

Also, this method would probably benefit from some line comments to explain the approach.

Reply:
done.

Reply:
(Mentioned the need for line comments in a new comment thread).

To clarify on 'done' here, does this mean you did verify that the returned value from Os.sysconf() is actually what we expect (i.e. the number of _actively_ running processors at that moment)?

Reply:
according to the different sources and the linux documentation, yes the returned value from Os.sysconf() is what we expect (online processors). I've added a documentation reference in the code as well for further validation.